application/src/main/java/org/thingsboard/server/controller/OtaPackageController.java

```diff
@@ -218,6 +181,26 @@ public OtaPackageInfo saveOtaPackageData(@ApiParam(value = OTA_PACKAGE_ID_PARAM_
         }
     }
 
+    private OtaPackage saveOtaPackageWithData(OtaPackageId otaPackageId, MultipartFile file, String checksum, ChecksumAlgorithm checksumAlgorithm) throws ThingsboardException, IOException {
+        OtaPackageInfo info = checkOtaPackageInfoId(otaPackageId, Operation.READ);
+        OtaPackage otaPackage = new OtaPackage(otaPackageId);
+        otaPackage.setCreatedTime(info.getCreatedTime());
+        otaPackage.setTenantId(info.getTenantId());
+        otaPackage.setDeviceProfileId(info.getDeviceProfileId());
+        otaPackage.setType(info.getType());
+        otaPackage.setTitle(info.getTitle());
+        otaPackage.setVersion(info.getVersion());
+        otaPackage.setTag(info.getTag());
+        otaPackage.setAdditionalInfo(info.getAdditionalInfo());
+        otaPackage.setChecksumAlgorithm(checksumAlgorithm);
+        otaPackage.setChecksum(checksum);
+        otaPackage.setFileName(file.getOriginalFilename());
+        otaPackage.setContentType(file.getContentType());
+        otaPackage.setDataSize(file.getSize());
+        otaPackage.setData(file.getInputStream());
+        return otaPackageService.saveOtaPackage(otaPackage);
```

Comment:
add MultipartFile as a second parameter to this method to avoid temporary file storage.

Reply:
research configuration parameters for the Tomcat temporary file storage and please provide link to those parameters